044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,342 @@
+similar disk usage between all brokers. Assuming all the brokers have similar disk usage levels, they will all apply
+rate limits at similar times and levels, effectively giving cluster-wide out-of-storage protection.
+
+This however provides limited protection to clusters with un-even distribution of partition replicas and thus storage
+usage. Additionally, it ties the plug-in directly to storage usage, there are other factors which users may wish to
+respond to.
+
+As clusters scale up the likelihood of even partition replica distribution drops. When partition replica are not evenly
+distributed it is possible for replication from broker `1` to broker `2` will cause broker `2` to consume all available
+storage without triggering throttling of clients, as broker `1` disk usage remains acceptable.
+
+Addressing the effects of uneven partition replicas distribution sounds like it should come
+under [KIP-73](https://cwiki.apache.org/confluence/display/KAFKA/KIP-73+Replication+Quotas). Unfortunately replication
+quotas are designed to manage the additional network load of migrating replicas between brokers, which does not address
+client publication leading to out of disk conditions through replication. As replication throttling is configured in
+terms of `TopicPartitions` the quota plug-in would need a mechanism to translate a logDir into the set of
+TopicPartitions replicated to that directory. Assuming it generated the appropriate throttle configuration this could
+lead to unpredictable latency spikes for producers configured with `acks = 1` or `acks = all` as the replication from
+partition leader to follower maybe delayed due to the throttle.
+
+Currently, the kafka-quotas-plugin considers the total quantity of storage attached to a single broker and how much of
+that is used ( see [issue#2](https://github.com/strimzi/kafka-quotas-plugin/issues/2)) when considering whether to apply
+throttling to clients. This is problematic with respect to handling disk failure for Just a Bunch Of Disks (JBOD)
+deployments [(KIP-112)](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD) as
+the broker will take partitions offline when the volume they are stored on runs out of space. Which in the case of
+unbalanced usage between volumes can lead to a volume running out of storage without throttling being applied.
+
+Throttling down message production on all broker nodes will protect the cluster from running out of disk due to replication.
+
+## Motivation
+
+Users need better protection against running out of disk on any of their nodes as it can degrade the service
+in unpredictable ways (corrupted segment logs) which could impede the interventions required to recover.
+
+## Proposal
+
+1. Extend the kafka-quotas-plugin so that we can choose to observe the disk usage 
+of all brokers in the cluster.
+2. [Deprecate](#compatibility) the existing local disk observations.
```

Comment:
I **strongly** vote for throwing all of the old stuff out of the window instead of deprecating it. We are only at version 0.2.0, it is only semi usable anyway and AFAIK nobody uses it. So we should make the live simple => no need to maintain it, test it and support it for another _years_.

Reply:
Sounds good, we'd prefer to remove the old code. Since we've got a couple of approvals, are you okay with us changing this @tombentley @Frawless

Reply:
I don't think it'll be too controversial given the current state of the plugin as you said @scholzj , we'll update the proposal.

Reply:
Have pushed up the change from keeping it backwards compatible to discarding the old implementation.